### ISSUE TITLE
pagination is not automatic

### ISSUE THREAD

**Issue by gabrielsroka** [okta/okta-sdk-java#787]
### Describe the bug?

1. The readme says that pagination is automatic, but doesn't provide a complete example. what does "automatic" mean? i don't have to write additional code?

i was able to get this working, but i don't know if it's correct. is there a better/shorter way? does the SDK do some of this for me? (i'm not a Java programmer.)
```java
import com.okta.sdk.client.Clients;
import com.okta.sdk.resource.common.PagedList;

import org.openapitools.client.ApiClient;
import org.openapitools.client.api.UserApi;
import org.openapitools.client.model.User;

class Main {
    public static void main(String[] args)  {
        ApiClient client = Clients.builder().build();
        UserApi userApi = new UserApi(client);
        String filter = "profile.lastName eq \"Doe\"";
        String after = null;
        do {
            PagedList<User> usersPagedList = 
                userApi.listUsersWithPaginationInfo(null, after, 2, filter, null, null, null);
            for (User user : usersPagedList.getItems()) {
                System.out.println(user.getProfile().getLogin());
            }
            after = getAfter(usersPagedList.getNextPage());
        } while (after != null);
    }

    public static String getAfter(String nextPage) {
        String after = null;
        if (nextPage != null) {
            String query = nextPage.split("\\?")[1];
            for (String pair : query.split("&")) {
                String[] nv = pair.split("=", 2);
                if (nv[0].equals("after")) after = nv[1];
            }
        }
        return after;
    }
}
```
https://replit.com/@gabrielsroka/okta-sdk-java#Main.java

parts of ^^ were stolen from
https://github.com/okta/okta-sdk-java/blob/cb0a648f39f2f7a7c72912d5c19bac0c44102f1e/integration-tests/src/test/groovy/com/okta/sdk/tests/it/UsersIT.groovy#L874-L880

2. Either way, the readme is wrong. `after` below should be the 2nd argument, not the 1st:

https://github.com/okta/okta-sdk-java/blob/cb0a648f39f2f7a7c72912d5c19bac0c44102f1e/README.md?plain=1#L423-L427

same with https://github.com/okta/okta-sdk-java/blob/cb0a648f39f2f7a7c72912d5c19bac0c44102f1e/examples/quickstart/src/main/java/quickstart/ReadmeSnippets.java#L282-L286

3. You can't use `q` to paginate users or groups, because the API doesn't return link headers when using `q`, so this:

```java
String q = "Doe";
PagedList<User> usersPagedList = userApi.listUsersWithPaginationInfo(q, after, 2, null, null, null, null);
```

returns this exception
```
Exception in thread "main" java.lang.NullPointerException:
  Cannot invoke "java.util.List.iterator()" because "linkHeaders" is null
    at org.openapitools.client.api.UserApi.listUsersWithPaginationInfo(UserApi.java:3506)
```

Note that apps, authorizationServers, idps, logs, etc. do return link headers with `q`, but users, groups, etc. don't.

4. This example doesn't work:

https://github.com/okta/okta-sdk-java/blob/cb0a648f39f2f7a7c72912d5c19bac0c44102f1e/README.md?plain=1#L184-L185

same with https://github.com/okta/okta-sdk-java/blob/cb0a648f39f2f7a7c72912d5c19bac0c44102f1e/examples/quickstart/src/main/java/quickstart/ReadmeSnippets.java#L103-L104

it'd need to be one of these:
```java
// q by email
List<User> users = userApi.listUsers("jcoder@example.com", null, 5, null, null, null, null);

// search by email
List<User> users = userApi.listUsers(null, null, 5, null, "profile.email eq \"jcoder@example.com\"", null, null);
```

### What is expected to happen?

automatic pagination

### What is the actual behavior?

it doesn't paginate. i had to write the code to figure out `after` and add the loop.

### Reproduction Steps?

see above

### Java Version

openjdk version "17.0.3"

### SDK Version

10.0

### OS version

NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

**Comment by gabrielsroka** [comment]
btw, where is the source code for `userApi.listUsersWithPaginationInfo` ? neither GH nor Google has ever seen it.

**Comment by gabrielsroka** [comment]
actually, ^^^ is really redundant, also, too, as well. 

why not this:
```java
HttpHeaders headers = responseEntity.getHeaders();
PagedList pagedList = new PagedList();
pagedList.addItems(Arrays.asList(responseEntity.getBody()));
List<String> linkHeaders = headers.get("link");
for (String link : linkHeaders) {
    String[] parts = link.split("; *");
    String url = parts[0]
        .replaceAll("<", "")
        .replaceAll(">", "");
    String rel = parts[1];
    if (rel.equals("rel=\"next\"")) {
        pagedList.setNextPage(url);
    } else if (rel.equals("rel=\"self\"")) {
        pagedList.setSelf(url);
    }
}
return pagedList;
```

again, i'm not a Java programmer, so there might be a better (or more idiomatic) way to write this.

**Comment by arvindkrishnakumar-okta** [comment]
@gabrielsroka thanks for posting! Apologies for the delayed response, will review this in detail and get back to you.

**Comment by gabrielsroka** [comment]
or, use a regex to parse the link header

```java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

// ...
String pattern = "<(.*)>; *rel=\"(.*)\"";
Pattern r = Pattern.compile(pattern);
// ...
for (String link : linkHeaders) {
    Matcher m = r.matcher(link);
    if (m.find()) {
        String url = m.group(1);
        String rel = m.group(2);
        if (rel.equals("next")) {
            pagedList.setNextPage(url);
        } else if (rel.equals("self")) {
            pagedList.setSelf(url);
        }
    }
}
```

**Comment by arvindkrishnakumar-okta** [comment]
@gabrielsroka Thanks for your suggestions! I did review the present pagination implementation and realized that the templates need to be reworked to omit redundant code. Your suggestions are noted and will be factored in.

**Comment by arvindkrishnakumar-okta** [comment]
@gabrielsroka merged PR to master, will cut a release soon. Please check it out and reopen this issue or open a new one if needed.

**Comment by arvindkrishnakumar-okta** [comment]
Released 10.2.0

**Comment by sdrissi** [comment]
I was reading through the doc as well and it is not clear if the SDK does go through all pages automatically. In the example of the doc it says "Loop through all of them". What does that mean? All of them in the first page or all of them as in all existing users?

```
UserApi userApi = new UserApi(client);
int limit = 2;
PagedList<User> pagedUserList = userApi.listUsersWithPaginationInfo(null, null, limit, null, null, null, null);

// loop through all of them
for (User tmpUser : pagedUserList.getItems()) {
    log.info("User: {}", tmpUser.getProfile().getEmail());
}

// or stream
pagedUserList.getItems().forEach(tmpUser -> log.info("User: {}", tmpUser.getProfile().getEmail()));
```

Let's say there are 10 existing users. Then, is the expected output two info logs (one for each of the 2 users in page 1 because limit is set to 2 users per page) or is it 10 info logs (i.e pagination is done automatically)?

**Comment by gabrielsroka** [comment]
@sdrissi my expectation/understanding is that it should give you all users. that said, it was broken in 10.0, but should be fixed in 10.2.

did you try it? (i have not)

@arvindkrishnakumar-okta , keep me honest here...

**Comment by sdrissi** [comment]
@gabrielsroka thanks for your reply. I did no try it yet. Would be great if @arvindkrishnakumar-okta could validate that this is the expected behavior. Will let you know once I try.
